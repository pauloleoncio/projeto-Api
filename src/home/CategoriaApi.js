

export function getAll() {
  return fetch('https://lojavirtual-api--tarleylana.repl.co/api/v1/categorias/')
            .then(res => res.json())
            .catch(err => {
              console.error('Falha ao tentar obter todas as categorias ' + err.message);
              throw new Error('NÃ£o foi possivel recuperar as categorias. Favor tentar novamente.')
            })
}

export function insert() {
  
}