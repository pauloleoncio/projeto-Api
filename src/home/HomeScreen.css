body {
  width: 1200px;
  margin: auto;
  font-family: 'Lucida Sans', 'Lucida Sans Regular', 'Lucida Grande', 'Lucida Sans Unicode', Geneva, Verdana, sans-serif
}

header img {
  width: 100px;
  height: 100px;
}

header {
  display: flex;
  background-color: #4CAF50;
  color: #ffffff;
  padding: 15px;
}

header h1 {
  flex-grow: 1;
  text-align: center;
}

section {
  display: flex;
}

section article {
  flex-basis: 40%;
  background-color: #32a4e7;
  color: white;
  padding: 20px;
}

section aside {
  display: flex;
  flex-grow: 1;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-around;
}

section aside div {
  height: 300px;
  width: 250px;
}

div img {
  width: 100px;
  height: 100px;
}

footer {
  border: 1px solid #d4d4d4;
  background-color: #f1f1f1;
  text-align: right;
  padding: 10px;
}