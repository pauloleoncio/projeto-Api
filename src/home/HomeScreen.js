import React, {useState, useEffect} from 'react';
import {Link, useHistory} from 'react-router-dom';

import Header from './Header';
import TextoSite from './TextoSite';
import CardCategoria from './CardCategoria';
import Footer from './Footer';

import './HomeScreen.css';

import {getAll} from './CategoriaApi';
/*
const valores = [
  {id: 1, nome: "Categoria 1", ativa: true},
  {id: 2, nome: "Categoria 2", ativa: true},
  {id: 3, nome: "Categoria 3", ativa: true},
  {id: 4, nome: "Categoria 4", ativa: true},
  {id: 5, nome: "Categoria 5", ativa: false},
  {id: 6, nome: "Categoria 6", ativa: true},
  {id: 7, nome: "Categoria 7", ativa: true},
  {id: 8, nome:"Categoria 8", ativa: true},
  {id: 9, nome:"Categoria 9", ativa: true}
]
*/

function HomeScreen() {
  const history = useHistory();

  const [loading, setLoading] = useState(true);
  const [categorias, setCategorias] = useState([]);
  const [msg, setMsg] = useState();
  
  useEffect(() => {
    const fetchData = async () => {
      try {
        const valores = await getAll();
        setCategorias(valores);
      } catch(err) {
        setMsg(err.message);
      } finally {
        setLoading(false);
      }
    }

    fetchData();
  }, [])

  return (
    <>
      <Header />
      <section data-testid="homescreen">
        <article>
          <TextoSite />
        </article>
        <aside>
            {loading && <h1>Carregando...</h1>}
            {msg && <h1>
                      {msg}
                      <button onClick={history.push('/')} >Atualizar</button>
                    </h1>}
            {
              categorias
                .filter(cat => cat.ativa === true)
                .map( cat => (
                  <Link key={cat.id} to={`/categorias/${cat.id}`}>
                    <CardCategoria
                      nome={cat.nome} 
                      ativo={cat.ativa}
                      img={cat.img128}
                    />
                  </Link>
              ))
          }
        </aside>
      </section>
      <Footer />
    </>
  )
}

export default HomeScreen;