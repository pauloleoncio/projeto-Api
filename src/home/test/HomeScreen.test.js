import React from 'react';
import { render, wait } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import HomeScreen from '../HomeScreen';
import * as CategoriaApi from '../CategoriaApi';

beforeAll(() => {
  CategoriaApi.getAll = jest.fn(() => responseOK);
});

test('Antes de carregar as categorias, o sistema deve exibir uma mensagem informando que está carregando', () => {
  
  let textLoading = null;
  
  act(() => {
    const component = render(<HomeScreen />, {wrapper: MemoryRouter});
    textLoading = component.getByText(/Carregando.../i);
  });

  expect(textLoading).toBeInTheDocument();
})

test('Após a execução da API, o sistema deve exibir as categorias', async () => {
  const component = render(<HomeScreen />, {wrapper: MemoryRouter});
  
  let categoriaAndroid = null;
  let categoriaIOS = null;
  let categoriaWindows = null;
  let categoriaLinux = null;
  await wait(() => {
    categoriaAndroid = component.getByText('Android');
    categoriaIOS = component.getByText('IOS');
    categoriaWindows = component.getByText('Windows');
    categoriaLinux = component.getByText('Linux');
  });
  
  expect(CategoriaApi.getAll).toHaveBeenCalled();

  expect(categoriaAndroid).toBeInTheDocument();
  expect(categoriaIOS).toBeInTheDocument();
  expect(categoriaWindows).toBeInTheDocument();
  expect(categoriaLinux).toBeInTheDocument();

  expect(component.getByTestId('homescreen')).toMatchSnapshot();
})

const responseOK = [
  {
    id: 1,
    nome: "Android",
    ativa: true,
    img64: 'https://lojavirtual-api--tarleylana.repl.co/img/android/android_64x64.png',
    img128: 'https://lojavirtual-api--tarleylana.repl.co/img/android/android_128x128.png',
    img256: 'https://lojavirtual-api--tarleylana.repl.co/img/android/android_256x256.png'
  },
  {
    id: 2,
    nome: "IOS",
    ativa: true,
    img64: 'https://lojavirtual-api--tarleylana.repl.co/img/ios/ios_64x64.png',
    img128: 'https://lojavirtual-api--tarleylana.repl.co/img/ios/ios_128x128.png',
    img256: 'https://lojavirtual-api--tarleylana.repl.co/img/ios/ios_256x256.png'
  },
  {
    id: 3,
    nome: "Windows",
    ativa: true,
    img64: 'https://lojavirtual-api--tarleylana.repl.co/img/windows/windows_64x64.png',
    img128: 'https://lojavirtual-api--tarleylana.repl.co/img/windows/windows_128x128.png',
    img256: 'https://lojavirtual-api--tarleylana.repl.co/img/windows/windows_256x256.png'
  },
  {
    id: 4,
    nome: "Linux",
    ativa: true,
    img64: 'https://lojavirtual-api--tarleylana.repl.co/img/linux/linux_64x64.png',
    img128: 'https://lojavirtual-api--tarleylana.repl.co/img/linux/linux_128x128.png',
    img256: 'https://lojavirtual-api--tarleylana.repl.co/img/linux/linux_256x256.png'
  }
]