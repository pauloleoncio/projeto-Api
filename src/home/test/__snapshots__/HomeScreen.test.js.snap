// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Após a execução da API, o sistema deve exibir as categorias 1`] = `
<section
  data-testid="homescreen"
>
  <article>
    <p>
      Texto que introduz o negócio e o serviço para o usuário e permite ao usuário navegar para qualquer uma das quatro categorias de produtos.
    </p>
  </article>
  <aside>
    <a
      href="/categorias/1"
    >
      <div>
        <h6>
          Android
        </h6>
        <img
          alt="Android"
          src="https://lojavirtual-api--tarleylana.repl.co/img/android/android_128x128.png"
        />
      </div>
    </a>
    <a
      href="/categorias/2"
    >
      <div>
        <h6>
          IOS
        </h6>
        <img
          alt="IOS"
          src="https://lojavirtual-api--tarleylana.repl.co/img/ios/ios_128x128.png"
        />
      </div>
    </a>
    <a
      href="/categorias/3"
    >
      <div>
        <h6>
          Windows
        </h6>
        <img
          alt="Windows"
          src="https://lojavirtual-api--tarleylana.repl.co/img/windows/windows_128x128.png"
        />
      </div>
    </a>
    <a
      href="/categorias/4"
    >
      <div>
        <h6>
          Linux
        </h6>
        <img
          alt="Linux"
          src="https://lojavirtual-api--tarleylana.repl.co/img/linux/linux_128x128.png"
        />
      </div>
    </a>
  </aside>
</section>
`;
